File: vscodeExt/215.kth-largest-element-in-an-array.py
#
# @lc app=leetcode id=215 lang=python3
#
# [215] Kth Largest Element in an Array
#

# @lc code=start
class Solution:

    def findKthLargest(self, nums: List[int], k: int) -> int:
        # New solution. construct a better particion. 
        return self.findKthParticion(nums,0, len(nums)-1, k)

    def findKthParticion(self, list, s, e, k):
        anchor = list[e]
        i = s-1 #[s, i]是 大于anchor的区间。 
        for j in range(s,e):
            if list[j] >= anchor:
                i = i+1 
                if i != j:
                    list[i], list[j] = list[j], list[i] # 小于anchor的向前移动。 
        list[i+1], list[e] = list[e], list[i+1]
        print(list[i+1], i+1)
        if i + 1 == k -1 :
            return list[i+1]
        if i + 1 > k -1:
            return self.findKthParticion( list, s, i, k)
        else:
            return self.findKthParticion(list, i+2, e, k)        
        # return i+1        
        # return -1

    def findKthLargest_1(self, nums: List[int], k: int) -> int:
        index = self.particion(nums, 0, len(nums)-1)
        # k-1 为相要的index 
        if index  == k -1:
            return nums[index]
        if index  < k -1:
            return self.findKthLargest(nums[index+1:], k-index -1)   
        else:
            return self.findKthLargest(nums[:index], k)     

    # sort list[s,e], return the index of anchor    
    def particion(self, list, s , e): #[start,end]是排序范围
        anchor = list[e]
        i = s-1 #[s, i]是 大于anchor的区间。 
        for j in range(s,e):
            if list[j] >= anchor:
                i = i+1 
                if i != j:
                    list[i], list[j] = list[j], list[i] # 小于anchor的向前移动。 
        list[i+1], list[e] = list[e], list[i+1]
        print(i+1)
        return i+1



# @lc code=end

File: vscodeExt/215.kth-largest-element-in-an-array.java
import java.util.Arrays;

/*
 * @lc app=leetcode id=215 lang=java
 *
 * [215] Kth Largest Element in an Array
 */

// @lc code=start
class Solution {
    public int findKthLargest(int[] nums, int k) {
        int index = partition(nums, 0, nums.length-1);
        if(index == k -1 ){
            return nums[index];
        }
        if(index < k -1){
            return findKthLargest(Arrays.copyOfRange(nums, index+1, nums.length), k-index -1);
        } else {
            return findKthLargest(Arrays.copyOfRange(nums, 0, index), k);
        }

        // return 0;
    }

    // return the right index of end;
    // find from [begin, end]
    public int partition(int[] nums, int begin, int end){
        if(begin < 0 || end >= nums.length){
            return -1;
        }
        int anchor = nums[end];
        // put every elem larger then anchor to the end;
        int place = end -1;
        for(int i = begin;i<=place;){
            //change the direction at here.
            if(nums[i]<=anchor){
                int t = nums[place];
                nums[place] = nums[i];
                nums[i] = t;
                place -=1;
            } else{
                i+=1;
            }
        }
        // swich placeholder and anchor;
        int t = nums[place+1];
        nums[place+1] = anchor;
        nums[end] = t;

        return place+1;
    }
}
// @lc code=end

